Add 'isCheckbox' prop to 'MenuItem'

Aljullu opened the pull request:
Fixes #1609.

- Adds a new prop `isCheckbox` to `<MenuItem>` that allows creating menu items with a `FormToggle` and [`role=menuitemcheckbox`](https://www.w3.org/TR/wai-aria-1.1/#menuitemcheckbox).
- Updates all menus that were using `<MenuItem>` with [`role=menuitem`](https://www.w3.org/TR/wai-aria-1.1/#menuitem) and had a checkbox inside that, to use the new prop.

### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader

### Screenshots
![image](https://user-images.githubusercontent.com/3616980/53014251-3b60cc80-3448-11e9-99f0-11133e5a161c.png)

### Detailed test instructions:
* Go to any report.
* Tab to columns menu (vertical ellipse).
* Verify only menu items are focusable but toggle buttons are not (before, you needed two Tabs to jump from one option to the next).
* Toggle a column off with the keyboard.
* Verify the column can be toggled back on with the keyboard.
* Repeat the same process with the _Dashboard_ menus: _Store Performance_, _Leaderboards_ and _Charts_.

Aljullu committed:
Make column filter menu items not focusable

Aljullu committed:
Check if onInvoke exists before calling it

jeffstieler commented:
I was unable to toggle the columns using the "enter" key.. 😕

psealock commented on packages/components/src/ellipsis-menu/menu-item.js:
```diff
@@ -20,26 +20,28 @@ class MenuItem extends Component {
-		if ( event.keyCode === ENTER || event.keyCode === SPACE ) {
+		const { onInvoke } = this.props;
+
+		if ( onInvoke && ( event.keyCode === ENTER || event.keyCode === SPACE ) ) {
```
Oddly, this isn't working on `ENTER` for me either.

joshuatf replied:
Looks like `onInvoke` is undefined so this condition isn't passing.

psealock commented on packages/components/src/table/index.js:
```diff
@@ -323,7 +323,7 @@ class TableCard extends Component {
 							}
 							return (
-								<MenuItem key={ key } onInvoke={ this.onColumnToggle( key ) }>
+								<MenuItem key={ key } tabIndex={ null }>
```
What does the `null` here do? Will this revert to the default prop? Is there a way to be more explicit (or remove altogether)?

psealock reviewed (commented):
Looking good and testing well. Just two comments

joshuatf reviewed (commented):
Tabbing is testing well.  Left a comment about `onInvoke` not being defined which seems to be why the enter key doesn't work.

Aljullu committed:
Revert "Check if onInvoke exists before calling it"

This reverts commit e3b63edd51395c0f967991e2d20fccf867e37b91.

Aljullu committed:
Revert "Make column filter menu items not focusable"

This reverts commit 3e0ac949fa057a0c235876731768b6f312ced998.

Aljullu committed:
Add isCheckbox prop to MenuItem component

Aljullu commented:
@jeffstieler @psealock @joshuatf thanks for the reviews! Good catch with the Enter key not working.

I reverted the previous commits and started this PR from scratch with a broader goal to fix the same issue also for the _Dashboard_ menus.

We were rendering checkboxes inside a `role=menuitem` element, which I think is not correct or, at least, not the best option here. Instead, with the new changes I converted those menu items to `role=menuitemcheckbox` and I added `aria-hidden` to the checkbox. That should fix #1609, while still allowing interaction with the Enter key.

I updated the intro post of this PR with a new title and test instructions.

jeffstieler commented:
I don't know if this needs to be debounced or what, but rapidly toggling twice causes anywhere from 2-4 toggles to occur:

![2019-02-20 10 23 26](https://user-images.githubusercontent.com/63922/53111262-03b25d00-34fa-11e9-99c5-500379e0e004.gif)

_(I realize it's hard to tell what's happening here - I'm toggling the control twice with either the space or enter key)_

ryelle reviewed (commented):
Testing with keyboard-only and with Safari + VoiceOver– works mostly as described, there's a glitch with arrow keys though. You should be able to arrow-down through the items, but there are two focus points per item, the menu item and then the toggle itself - and when focused on the toggle, <kbd>enter</kbd> doesn't trigger the action. I'm assuming this has something to do with how the `NavigableMenu` manages key focus. Not sure this is enough of a bug to prevent merging this, but thought I'd mention in case it's an easy fix :)

Aljullu committed:
Fix arrow keys navigation in checkbox menu items

Aljullu committed:
Merge branch 'master' into fix/column-filter-toggle

Aljullu commented:
@ryelle thanks for testing it! Arrow keys navigation should be fixed in e240d466a8d26029de58231bef06cde8d6428d0c.

> I don't know if this needs to be debounced or what, but rapidly toggling twice causes anywhere from 2-4 toggles to occur:

@jeffstieler I'm not sure to understand what the problem is here. Do you mean we should disable the toggle buttons while it's updating?

joshuatf commented:
>I don't know if this needs to be debounced or what, but rapidly toggling twice causes anywhere from 2-4 toggles to occur:

This is caused by updating the state and the subsequent (slightly delayed) update that's received when the props are modified after a successful API request.

Since we're loading all settings into the initial state, we can probably safely skip updating the state within `componentDidUpdate()` and only modify the state if an error is received.  That might be worth opening a separate issue for; I don't mind outlining the problem and/or handling that if that's the direction we want to go.

/cc @jeffstieler @Aljullu

Aljullu committed:
Table: don't update showCols state when the headers prop changes

Aljullu commented:
Good point, I changed that behavior in the last commit (2dd093da7c4d3e55b0101d8d211c379c90596c75), so `<Table>` doesn't update the `showCols` state when the `headers` prop changes.

jeffstieler commented:
Looks like the rapid toggling issue was fixed with 2dd093d 👍

timmyc commented:
It seems all the issues found in review have either been handled, or follow-up issues opened. As such I'm marking this one as ready to merge.

timmyc reviewed (approved):
:shipit:
